torchao/csrc/metal/int4_quantized_kernels.metal

```diff
@@ -0,0 +1,128 @@
+#include <metal_simdgroup>
+#include <metal_stdlib>
+using namespace metal;
+
+template <typename T> struct Vec4Type {};
+
+template <> struct Vec4Type<float> { using type = float4; };
+
+template <> struct Vec4Type<half> { using type = half4; };
+
+#if __METAL_VERSION__ >= 310
+template <> struct Vec4Type<bfloat> { using type = bfloat4; };
+#endif
+
+template <typename T, unsigned groupSize>
+kernel void int4pack_mv(constant T *A [[buffer(0)]],
```

Comment:
can you please add comment(s) to clarify what this kernel is doing along the different quantization configuration knobs available? I shouldn't have to read and deeply understand the implementation of the function to figure out what it does.

- clearly 4-bit
- either rename A/B to make it clear which is the matrix and which is the vector (and specify M/K/N meaning in relation to these matrices) or exactly follow gemv in all aspects of the interface (looks like the order is vector/matrix despite the function name being `mv`); my preference would be clarity but I would be willing to defer to decades of precedent with the BLAS interface
- I see that it has groupSize, but it probably wouldn't hurt to spell out that it's groupwise quantization while we are here
- symmetric or asymmetric quantization for each of weights and activations?
- is the matrix column-major or row-major? (is also discussing transposition in addition to that relevant?)

Reply:
> either rename A/B to make it clear which is the matrix and which is the vector (and specify M/K/N meaning in relation to these matrices) or exactly follow gemv in all aspects of the interface (looks like the order is vector/matrix despite the function name being mv); my preference would be clarity but I would be willing to defer to decades of precedent with the BLAS interface

I dont intend to stick with BLAS interface yet. Especially if we have to do any fancy packing. Will update the kernel name to something else.

> I see that it has groupSize, but it probably wouldn't hurt to spell out that it's groupwise quantization while we are here

Not sure if it helps readability significantly.

> symmetric or asymmetric quantization for each of weights and activations?

THis is weight only quant as of now.

I think the way we probably will handle symmetric vs asymmetirc is via same code handling zero point addition or now via macros or template param.

> is the matrix column-major or row-major? (is also discussing transposition in addition to that relevant?)

Comments about the args will address this